Skip allreducing `local_used_maps_dev_` when `find_unused_param=False`

sinannasir opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40407 Skip allreducing `local_used_maps_dev_` when `find_unused_param=False`**

Summary: 
1. In reducer.cpp, we have a new boolean `find_unused_param_` and its value is set in `Reducer::prepare_for_backward`.
If `!find_unused_param_`, then it avoids `allreduce(local_used_maps_dev_)`.
2. Solves issue [38942](https://github.com/pytorch/pytorch/issues/38942).
3. Fixes incorrect `find_unused_parameters_` passing like checking `outputs.empty()` or `unused_parameters_.empty()`.

Test Plan: 
1. Run `test/distributed/test_c10d.py` and make sure all tests pass.
2. A new test case `test_find_unused_parameters_when_unused_parameters_empty` is included. Old `reducer.cpp` was failing in that unit test because it was checking `find_unused_parameters_` by `unused_parameters_.empty()`. Current `reducer.cpp` passes this unit test.
3. Two test cases were failing `test_forward_backward_unused_parameters` and `test_forward_backward_optimizer` , because `find_unused_parameter_` of their `reducer` object was not set properly. I fixed that as well.

Tasks: T68705534

Tags: DDP

Differential Revision: [D22176231](https://www.internalfb.com/intern/diff/D22176231/)

sinannasir committed:
Skip allreducing `local_used_maps_dev_` when `find_unused_param=False`

Summary: Solve issue https://github.com/pytorch/pytorch/issues/38942.
In reducer.cpp, I check whether `find_unused_param` is set to False by !unused_parameters_.empty().
If !unused_parameters_.empty(), then it avoids `allreduce(local_used_maps_dev_)`.

Test Plan: Run `test/distributed/test_c10d.py` and make sure all tests pass.

Reviewers: Pritam Damania

Subscribers: Pritam Damania, Shen Li

Tasks: T68705534

Tags: DDP

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -392,13 +392,15 @@ void Reducer::mark_variable_ready(VariableIndex index) {
-      // We do async H2D to avoid the blocking overhead. The async copy and
-      // allreduce respect the current stream, so will be sequenced correctly.
-      local_used_maps_dev_[i].copy_(local_used_maps_[i], true);
+    if (!unused_parameters_.empty()) {
```
This does not guarantee to be consistent across processes. Say all processes set `find_unused_parameters = true`, but some processes have unused parameters but other do not. In that case, some processes will never participating in this AllReduce, leading to either hang or wrong results. 

To know it deterministically, you might want to directly passing `find_unused_parameters` from Python to C++.

zhaojuanmao replied:
also, does not need to allocate memory for local_used_maps_dev_ and local_used_maps if find_unused_parameters is false

sinannasir replied:
@zhaojuanmao Do you recommend to skip all use of `local_used_maps_dev_` and `local_used_maps` when `find_unused_param=False`? I think `local_used_maps_dev_` can be abandoned completely when `find_unused_param=False`, but `local_used_maps_` is still required in `Reducer::finalize_bucket_dense`.

zhaojuanmao replied:
@sinannasir we can still keep the variables, but not allocate memory to them

mrshenli commented:
And since the unused parameters from different processes can be different, let's also add a test for that case.

sinannasir committed:
Update on "Skip allreducing `local_used_maps_dev_` when `find_unused_param=False`"


Summary: 
1. Solves issue [38942](https://github.com/pytorch/pytorch/issues/38942).
2. In reducer.cpp, I check whether `find_unused_param` is set to `false` by `!unused_parameters_.empty()`.
If `!unused_parameters_.empty()`, then it avoids `allreduce(local_used_maps_dev_)`.
3. `unused_parameters_()` is always empty, as we set `find_unused_param = false`. Further, there may be additional cases with an empty `unused_parameters_`, but `find_unused_param = true`. Therefore, in addition to the issue [38942](https://github.com/pytorch/pytorch/issues/38942), this diff also avoids `allreduce(local_used_maps_dev_)` for that case.
I think that since we never call `unused_parameters_.clear()` in separate processes, this will not cause any issue and will not lead to DDP comm hang. In fact, it will be more beneficial, because if all parameters are used, then each should already be reduced. 

Test Plan: Run `test/distributed/test_c10d.py` and make sure all tests pass.

Reviewers: Pritam Damania

Subscribers: Pritam Damania, Shen Li

Tasks: T68705534

Tags: DDP

Differential Revision: [D22176231](https://www.internalfb.com/intern/diff/D22176231/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -752,8 +755,10 @@ void Reducer::prepare_for_backward(
   // variables will be called, and we don't have to search the autograd graph
   // for presence of these hooks.
   if (outputs.empty()) {
+    find_unused_parameters_ = false;
```
This does not guarantee to be correct either. What if one process's forward pass produces empty outputs but still sets `find_unused_parameters_ =rure`?

sinannasir replied:
Okay, will move it to the constructor and also address @zhaojuanmao 's comment.

sinannasir committed:
Update on "Skip allreducing `local_used_maps_dev_` when `find_unused_param=False`"


Summary: 
1. Solves issue [38942](https://github.com/pytorch/pytorch/issues/38942).
2. In reducer.cpp, we have a new boolean `find_unused_param_` and its value is set in `Reducer::prepare_for_backward`.
If `!find_unused_param_`, then it avoids `allreduce(local_used_maps_dev_)`.

Test Plan: Run `test/distributed/test_c10d.py` and make sure all tests pass.

Reviewers: Pritam Damania

Subscribers: Pritam Damania, Shen Li, Yanli Zhao

Tasks: T68705534

Tags: DDP

Differential Revision: [D22176231](https://www.internalfb.com/intern/diff/D22176231/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -126,33 +128,34 @@ Reducer::Reducer(
         backward_stats_.end(),
         [=](std::vector<int64_t>& v) { v.resize(variable_count); });
   }
+  if (find_unused_parameters_) {
```
Can we add a note explaining why we can avoid initializing these variables when `find_unused_parameters_` is false? We can refer to that note in each place we have such logic. Example of note: https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/engine.cpp#L100. Example of referring to a note: https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/engine.cpp#L64.

sinannasir replied:
There were two errors caused by setting `global_unused=false` when `find_unused_parameters_ = false`. I keep using `local_used_maps_` when `find_unused_parameters_ = false` to avoid the errors and now it passes all the cases. Let me know about the alternatives.
Errors were as follows:
```
..............ERROR:root:Caught exception: 
Traceback (most recent call last):
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 207, in wrapper
    fn()
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 76, in wrapper
    return func(*args, **kwargs)
  File "test/distributed/test_c10d.py", line 2380, in test_global_local_unused_params_grad
    run_and_verify_grad(cpu_model)
  File "test/distributed/test_c10d.py", line 2369, in run_and_verify_grad
    self.assertIsNone(task_unused_p.grad)
  File "/home/sinannasir/local/miniconda3/envs/pytorch/lib/python3.8/unittest/case.py", line 1322, in assertIsNone
    self.fail(self._formatMessage(msg, standardMsg))
  File "/home/sinannasir/local/miniconda3/envs/pytorch/lib/python3.8/unittest/case.py", line 753, in fail
    raise self.failureException(msg)
AssertionError: tensor([[0., 0.],
        [0., 0.]]) is not None
exiting process with exit code: 10
ERROR:root:Caught exception: 
Traceback (most recent call last):
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 207, in wrapper
    fn()
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 76, in wrapper
    return func(*args, **kwargs)
  File "test/distributed/test_c10d.py", line 2380, in test_global_local_unused_params_grad
    run_and_verify_grad(cpu_model)
  File "test/distributed/test_c10d.py", line 2369, in run_and_verify_grad
    self.assertIsNone(task_unused_p.grad)
  File "/home/sinannasir/local/miniconda3/envs/pytorch/lib/python3.8/unittest/case.py", line 1322, in assertIsNone
    self.fail(self._formatMessage(msg, standardMsg))
  File "/home/sinannasir/local/miniconda3/envs/pytorch/lib/python3.8/unittest/case.py", line 753, in fail
    raise self.failureException(msg)
AssertionError: tensor([[0., 0.],
        [0., 0.]]) is not None
exiting process with exit code: 10
Process 0 terminated with exit code 10, terminating remaining processes.
Es.....s...............................................test/distributed/test_c10d.py:1531: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  tensor = torch.full([100, 100], self.rank)
test/distributed/test_c10d.py:1531: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  tensor = torch.full([100, 100], self.rank)
test/distributed/test_c10d.py:1531: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  tensor = torch.full([100, 100], self.rank)
test/distributed/test_c10d.py:1531: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  tensor = torch.full([100, 100], self.rank)
.test/distributed/test_c10d.py:1554: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  self.assertEqual(torch.full([10, 10], self.world_size), tensor)
test/distributed/test_c10d.py:1554: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  self.assertEqual(torch.full([10, 10], self.world_size), tensor)
test/distributed/test_c10d.py:1554: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  self.assertEqual(torch.full([10, 10], self.world_size), tensor)
test/distributed/test_c10d.py:1554: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  self.assertEqual(torch.full([10, 10], self.world_size), tensor)
.....s...............F...............
======================================================================
ERROR: test_global_local_unused_params_grad (__main__.DistributedDataParallelTest)
By simulating a multi-task training, this test is to make sure:
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 204, in wrapper
    self._join_processes(fn)
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 306, in _join_processes
    self._check_return_codes(elapsed_time)
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 339, in _check_return_codes
    raise RuntimeError(error)
RuntimeError: Processes 0 1 exited with error code 10

======================================================================
FAIL: test_forward_backward_unused_parameters (__main__.ReducerTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/distributed/test_c10d.py", line 3049, in test_forward_backward_unused_parameters
    self.assertEqual(None, model.fc3.weight.grad)
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_utils.py", line 1124, in assertEqual
    super().assertEqual(x, y, msg=msg)
AssertionError: None != tensor([[0., 0., 0., 0.],
        [0., 0.[58 chars]0.]])

----------------------------------------------------------------------
Ran 107 tests in 175.165s

FAILED (failures=1, errors=1, skipped=3)
```

mrshenli replied:
hmm, I checked `test_global_local_unused_params_grad`, and looks like it is setting `find_unused_parameters` to true. So in this case `global_unused` should not be directly set to `false` no?

mrshenli replied:
For `test_forward_backward_unused_parameters`, if you look at how the reducer is created, it does not pass `find_unused_parameters` properly, and relies on `prepare_for_backward` to detect that using the forward `output`. So you will need to modify the test accordingly.

sinannasir replied:
> For `test_forward_backward_unused_parameters`, if you look at how the reducer is created, it does not pass `find_unused_parameters` properly, and relies on `prepare_for_backward` to detect that using the forward `output`. So you will need to modify the test accordingly.

I see your point. For those cases, it was setting `find_unused_parameters_` to False by default and not passing the actual value. I will revert reducer.cpp back and keep avoiding initializing `local_used_parameters_`, and look into the test cases. Thanks,

sinannasir replied:
> hmm, I checked `test_global_local_unused_params_grad`, and looks like it is setting `find_unused_parameters` to true. So in this case `global_unused` should not be directly set to `false` no?

I was declaring global_unused twice when `find_unused_parameters=true`. I fixed that and now all test cases pass without any modification on `test_forward_backward_unused_parameters`. I still modified `test_forward_backward_unused_parameters` and pass `find_unused_parameters` properly for correctness.

sinannasir replied:
> Can we add a note explaining why we can avoid initializing these variables when `find_unused_parameters_` is false? We can refer to that note in each place we have such logic. Example of note: https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/engine.cpp#L100. Example of referring to a note: https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/engine.cpp#L64.

Working on this feedback.

pritamdamania87 reviewed (commented):
> And since the unused parameters from different processes can be different, let's also add a test for that case.

@mrshenli Could you elaborate on the test case we need here? Looks like we have unit tests where we test both cases for `find_unused_parameters_`. Is there something missing in our current tests?

sinannasir committed:
Update on "Skip allreducing `local_used_maps_dev_` when `find_unused_param=False`"


Summary: 
1. Solves issue [38942](https://github.com/pytorch/pytorch/issues/38942).
2. In reducer.cpp, we have a new boolean `find_unused_param_` and its value is set in `Reducer::prepare_for_backward`.
If `!find_unused_param_`, then it avoids `allreduce(local_used_maps_dev_)`.

Test Plan: Run `test/distributed/test_c10d.py` and make sure all tests pass.

Reviewers: Pritam Damania

Subscribers: Pritam Damania, Shen Li, Yanli Zhao

Tasks: T68705534

Tags: DDP

Differential Revision: [D22176231](https://www.internalfb.com/intern/diff/D22176231/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -147,12 +148,15 @@ Reducer::Reducer(
-      local_used_maps_dev_[i] =
-          at::empty({static_cast<long>(variable_count)}, options);
+      if (find_unused_parameters_) {
+        local_used_maps_dev_.resize(replica_count);
```
should the resize happen before the for loop? Is there any reason that we need this in every iteration?

sinannasir replied:
> should the resize happen before the for loop? Is there any reason that we need this in every iteration?

Right, I will fix it.

mrshenli commented:
> @mrshenli Could you elaborate on the test case we need here? Looks like we have unit tests where we test both cases for find_unused_parameters_. Is there something missing in our current tests?

Sure. The first two attempts in this PR reminds us that we might want to add tests for the following cases:

1. `unused_parameters_.empty()` does not imply `find_unused_parameters`: we can set `find_unused_parameters` to `true`, and then let some some process use all parameters but other don't.
2. `outputs` in `prepare_for_backward` does not imply `find_unused_parameters`: the code today actually implies `find_unused_parameters` using `outputs`. However, as we know explicitly passing `find_unused_parameters`, we can modify the following code accordingly and add tests for it, where the `forward` pass in some process produces empty outputs.:
     
 https://github.com/pytorch/pytorch/blob/e490352dc4ad4f61c795bfdae968954c7d916861/torch/csrc/distributed/c10d/reducer.cpp#L751-L756

sinannasir committed:
Update on "Skip allreducing `local_used_maps_dev_` when `find_unused_param=False`"


Summary: 
1. Solves issue [38942](https://github.com/pytorch/pytorch/issues/38942).
2. In reducer.cpp, we have a new boolean `find_unused_param_` and its value is set in `Reducer::prepare_for_backward`.
If `!find_unused_param_`, then it avoids `allreduce(local_used_maps_dev_)`.

Test Plan: Run `test/distributed/test_c10d.py` and make sure all tests pass.

Reviewers: Pritam Damania

Subscribers: Pritam Damania, Shen Li, Yanli Zhao

Tasks: T68705534

Tags: DDP

Differential Revision: [D22176231](https://www.internalfb.com/intern/diff/D22176231/)

[ghstack-poisoned]

sinannasir commented:
There were two errors caused by setting global_unused=false when find_unused_parameters_ = false. I keep using local_used_maps_ when find_unused_parameters_ = false to avoid the errors and now it passes all the cases. Let me know about the alternatives.
Errors were as follows:

..............ERROR:root:Caught exception: 
Traceback (most recent call last):
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 207, in wrapper
    fn()
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 76, in wrapper
    return func(*args, **kwargs)
  File "test/distributed/test_c10d.py", line 2380, in test_global_local_unused_params_grad
    run_and_verify_grad(cpu_model)
  File "test/distributed/test_c10d.py", line 2369, in run_and_verify_grad
    self.assertIsNone(task_unused_p.grad)
  File "/home/sinannasir/local/miniconda3/envs/pytorch/lib/python3.8/unittest/case.py", line 1322, in assertIsNone
    self.fail(self._formatMessage(msg, standardMsg))
  File "/home/sinannasir/local/miniconda3/envs/pytorch/lib/python3.8/unittest/case.py", line 753, in fail
    raise self.failureException(msg)
AssertionError: tensor([[0., 0.],
        [0., 0.]]) is not None
exiting process with exit code: 10
ERROR:root:Caught exception: 
Traceback (most recent call last):
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 207, in wrapper
    fn()
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 76, in wrapper
    return func(*args, **kwargs)
  File "test/distributed/test_c10d.py", line 2380, in test_global_local_unused_params_grad
    run_and_verify_grad(cpu_model)
  File "test/distributed/test_c10d.py", line 2369, in run_and_verify_grad
    self.assertIsNone(task_unused_p.grad)
  File "/home/sinannasir/local/miniconda3/envs/pytorch/lib/python3.8/unittest/case.py", line 1322, in assertIsNone
    self.fail(self._formatMessage(msg, standardMsg))
  File "/home/sinannasir/local/miniconda3/envs/pytorch/lib/python3.8/unittest/case.py", line 753, in fail
    raise self.failureException(msg)
AssertionError: tensor([[0., 0.],
        [0., 0.]]) is not None
exiting process with exit code: 10
Process 0 terminated with exit code 10, terminating remaining processes.
Es.....s...............................................test/distributed/test_c10d.py:1531: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  tensor = torch.full([100, 100], self.rank)
test/distributed/test_c10d.py:1531: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  tensor = torch.full([100, 100], self.rank)
test/distributed/test_c10d.py:1531: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  tensor = torch.full([100, 100], self.rank)
test/distributed/test_c10d.py:1531: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  tensor = torch.full([100, 100], self.rank)
.test/distributed/test_c10d.py:1554: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  self.assertEqual(torch.full([10, 10], self.world_size), tensor)
test/distributed/test_c10d.py:1554: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  self.assertEqual(torch.full([10, 10], self.world_size), tensor)
test/distributed/test_c10d.py:1554: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  self.assertEqual(torch.full([10, 10], self.world_size), tensor)
test/distributed/test_c10d.py:1554: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  self.assertEqual(torch.full([10, 10], self.world_size), tensor)
.....s...............F...............
======================================================================
ERROR: test_global_local_unused_params_grad (__main__.DistributedDataParallelTest)
By simulating a multi-task training, this test is to make sure:
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 204, in wrapper
    self._join_processes(fn)
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 306, in _join_processes
    self._check_return_codes(elapsed_time)
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 339, in _check_return_codes
    raise RuntimeError(error)
RuntimeError: Processes 0 1 exited with error code 10

======================================================================
FAIL: test_forward_backward_unused_parameters (__main__.ReducerTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/distributed/test_c10d.py", line 3049, in test_forward_backward_unused_parameters
    self.assertEqual(None, model.fc3.weight.grad)
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_utils.py", line 1124, in assertEqual
    super().assertEqual(x, y, msg=msg)
AssertionError: None != tensor([[0., 0., 0., 0.],
        [0., 0.[58 chars]0.]])

----------------------------------------------------------------------
Ran 107 tests in 175.165s

FAILED (failures=1, errors=1, skipped=3)

sinannasir commented:
There were two errors caused by setting `global_unused=false` when `find_unused_parameters_ = false`. I keep using `local_used_maps_` when `find_unused_parameters_ = false` to avoid the errors and now it passes all the cases. Let me know about the alternatives.
Errors were as follows:
```
..............ERROR:root:Caught exception: 
Traceback (most recent call last):
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 207, in wrapper
    fn()
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 76, in wrapper
    return func(*args, **kwargs)
  File "test/distributed/test_c10d.py", line 2380, in test_global_local_unused_params_grad
    run_and_verify_grad(cpu_model)
  File "test/distributed/test_c10d.py", line 2369, in run_and_verify_grad
    self.assertIsNone(task_unused_p.grad)
  File "/home/sinannasir/local/miniconda3/envs/pytorch/lib/python3.8/unittest/case.py", line 1322, in assertIsNone
    self.fail(self._formatMessage(msg, standardMsg))
  File "/home/sinannasir/local/miniconda3/envs/pytorch/lib/python3.8/unittest/case.py", line 753, in fail
    raise self.failureException(msg)
AssertionError: tensor([[0., 0.],
        [0., 0.]]) is not None
exiting process with exit code: 10
ERROR:root:Caught exception: 
Traceback (most recent call last):
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 207, in wrapper
    fn()
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 76, in wrapper
    return func(*args, **kwargs)
  File "test/distributed/test_c10d.py", line 2380, in test_global_local_unused_params_grad
    run_and_verify_grad(cpu_model)
  File "test/distributed/test_c10d.py", line 2369, in run_and_verify_grad
    self.assertIsNone(task_unused_p.grad)
  File "/home/sinannasir/local/miniconda3/envs/pytorch/lib/python3.8/unittest/case.py", line 1322, in assertIsNone
    self.fail(self._formatMessage(msg, standardMsg))
  File "/home/sinannasir/local/miniconda3/envs/pytorch/lib/python3.8/unittest/case.py", line 753, in fail
    raise self.failureException(msg)
AssertionError: tensor([[0., 0.],
        [0., 0.]]) is not None
exiting process with exit code: 10
Process 0 terminated with exit code 10, terminating remaining processes.
Es.....s...............................................test/distributed/test_c10d.py:1531: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  tensor = torch.full([100, 100], self.rank)
test/distributed/test_c10d.py:1531: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  tensor = torch.full([100, 100], self.rank)
test/distributed/test_c10d.py:1531: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  tensor = torch.full([100, 100], self.rank)
test/distributed/test_c10d.py:1531: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  tensor = torch.full([100, 100], self.rank)
.test/distributed/test_c10d.py:1554: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  self.assertEqual(torch.full([10, 10], self.world_size), tensor)
test/distributed/test_c10d.py:1554: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  self.assertEqual(torch.full([10, 10], self.world_size), tensor)
test/distributed/test_c10d.py:1554: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  self.assertEqual(torch.full([10, 10], self.world_size), tensor)
test/distributed/test_c10d.py:1554: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  self.assertEqual(torch.full([10, 10], self.world_size), tensor)
.....s...............F...............
======================================================================
ERROR: test_global_local_unused_params_grad (__main__.DistributedDataParallelTest)
By simulating a multi-task training, this test is to make sure:
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 204, in wrapper
    self._join_processes(fn)
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 306, in _join_processes
    self._check_return_codes(elapsed_time)
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 339, in _check_return_codes
    raise RuntimeError(error)
RuntimeError: Processes 0 1 exited with error code 10

======================================================================
FAIL: test_forward_backward_unused_parameters (__main__.ReducerTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/distributed/test_c10d.py", line 3049, in test_forward_backward_unused_parameters
    self.assertEqual(None, model.fc3.weight.grad)
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_utils.py", line 1124, in assertEqual
    super().assertEqual(x, y, msg=msg)
AssertionError: None != tensor([[0., 0., 0., 0.],
        [0., 0.[58 chars]0.]])

----------------------------------------------------------------------
Ran 107 tests in 175.165s

FAILED (failures=1, errors=1, skipped=3)
```

sinannasir committed:
Update on "Skip allreducing `local_used_maps_dev_` when `find_unused_param=False`"


Summary: 
1. Solves issue [38942](https://github.com/pytorch/pytorch/issues/38942).
2. In reducer.cpp, we have a new boolean `find_unused_param_` and its value is set in `Reducer::prepare_for_backward`.
If `!find_unused_param_`, then it avoids `allreduce(local_used_maps_dev_)`.

Test Plan: Run `test/distributed/test_c10d.py` and make sure all tests pass.

Reviewers: Pritam Damania

Subscribers: Pritam Damania, Shen Li, Yanli Zhao

Tasks: T68705534

Tags: DDP

Differential Revision: [D22176231](https://www.internalfb.com/intern/diff/D22176231/)

[ghstack-poisoned]

sinannasir commented:
1. All tests pass and we don't allocate memory for `local_used_maps` and `local_used_maps_dev_` if find_unused_parameters is false. 
2. Fixed declaring `global_unused` twice when `find_unused_parameters_ = true`.
 3. Correctly passing `find_unused_parameters` for test case `test_forward_backward_unused_parameters`. 
4. Included *Note [Skip allreducing local_used_maps_dev_]* for reference.

sinannasir committed:
Update on "Skip allreducing `local_used_maps_dev_` when `find_unused_param=False`"


Summary: 
1. Solves issue [38942](https://github.com/pytorch/pytorch/issues/38942).
2. In reducer.cpp, we have a new boolean `find_unused_param_` and its value is set in `Reducer::prepare_for_backward`.
If `!find_unused_param_`, then it avoids `allreduce(local_used_maps_dev_)`.

Test Plan: Run `test/distributed/test_c10d.py` and make sure all tests pass.

Reviewers: Pritam Damania

Subscribers: Pritam Damania, Shen Li, Yanli Zhao

Tasks: T68705534

Tags: DDP

Differential Revision: [D22176231](https://www.internalfb.com/intern/diff/D22176231/)

[ghstack-poisoned]

sinannasir commented:
Avoided some VS Code editorial changes.

sinannasir commented:
> > @mrshenli Could you elaborate on the test case we need here? Looks like we have unit tests where we test both cases for find_unused_parameters_. Is there something missing in our current tests?
> 
> Sure. The first two attempts in this PR reminds us that we might want to add tests for the following cases:
> 
> 1. `unused_parameters_.empty()` does not imply `find_unused_parameters`: we can set `find_unused_parameters` to `true`, and then let some some process use all parameters but other don't.
> 2. `outputs` in `prepare_for_backward` does not imply `find_unused_parameters`: the code today actually implies `find_unused_parameters` using `outputs`. However, as we know explicitly passing `find_unused_parameters`, we can modify the following code accordingly and add tests for it, where the `forward` pass in some process produces empty outputs.:
> 
> https://github.com/pytorch/pytorch/blob/e490352dc4ad4f61c795bfdae968954c7d916861/torch/csrc/distributed/c10d/reducer.cpp#L751-L756

I am looking into it.

sinannasir committed:
Update on "Skip allreducing `local_used_maps_dev_` when `find_unused_param=False`"


Summary: 
1. Solves issue [38942](https://github.com/pytorch/pytorch/issues/38942).
2. In reducer.cpp, we have a new boolean `find_unused_param_` and its value is set in `Reducer::prepare_for_backward`.
If `!find_unused_param_`, then it avoids `allreduce(local_used_maps_dev_)`.

Test Plan: Run `test/distributed/test_c10d.py` and make sure all tests pass.

Reviewers: Pritam Damania

Subscribers: Pritam Damania, Shen Li, Yanli Zhao

Tasks: T68705534

Tags: DDP

Differential Revision: [D22176231](https://www.internalfb.com/intern/diff/D22176231/)

[ghstack-poisoned]

sinannasir commented:
The new changes in the final commit:
1. A new test case `test_find_unused_parameters_when_unused_parameters_empty` is included. Today's `reducer.cpp` is failing in that unit test because it is checking `find_unused_parameters_` by using unused_parameters_:
https://github.com/pytorch/pytorch/blob/e490352dc4ad4f61c795bfdae968954c7d916861/torch/csrc/distributed/c10d/reducer.cpp#L380-L396

The problem is with `backward`, but the exception is happening in the constructor of the next ddp_model. This is because the broadcast `process_group_->broadcast(vec)->wait();` starts to pass weird `control_accessor` entries after the first problematic `backward` call. This may have something to do with `gloo` process group. The test now passes after new modifications.
2. Writing a test scenario for the following statement is not possible because of #40566 
> outputs in prepare_for_backward does not imply find_unused_parameters: the code today actually implies find_unused_parameters using outputs.
I still changed the way `find_parameters_unused_` is checked in the `prepare_for_backward` function. 

3. Two test cases were failing `test_forward_backward_unused_parameters` and `test_forward_backward_optimizer` , because `find_unused_parameter_` of their `reducer` object was not set properly. I fixed that as well.

pritamdamania87 commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -369,26 +385,27 @@ void Reducer::autograd_hook(VariableIndex index) {
     rebuilt_params_.push_back(
         replicas_[index.replica_index][index.variable_index]);
     rebuilt_param_indices_.push_back(index.variable_index);
   }
 
-  // If there are model parameters that went unused when computing the model
-  // output, they won't be part of the autograd graph, and won't receive
-  // gradients. These parameters are discovered in the `prepare_for_backward`
-  // function and their indexes stored in the `unused_parameters_` vector.
-  if (!has_marked_unused_parameters_ && !unused_parameters_.empty()) {
+  // If `find_unused_parameters_` is true there may be model parameters that
+  // went unused when computing the model output, they won't be part of the
+  // autograd graph, and won't receive gradients. These parameters are discovered
+  // in the `prepare_for_backward` function and their indexes stored in
+  // the `unused_parameters_` vector.
+  if (!has_marked_unused_parameters_ && find_unused_parameters_) {
```
These changes look good to me, although I'll let Shen stamp the PR.

sinannasir replied:
Thanks for letting me know about the merge conflict :) Just rebased and resolved the conflict at `init.cpp`.

pritamdamania87 reviewed (commented):
Looks like there are some merge conflicts in CI you probably need to rebase and resubmit the PR.

sinannasir committed:
Update on "Skip allreducing `local_used_maps_dev_` when `find_unused_param=False`"


Summary: 
1. In reducer.cpp, we have a new boolean `find_unused_param_` and its value is set in `Reducer::prepare_for_backward`.
If `!find_unused_param_`, then it avoids `allreduce(local_used_maps_dev_)`.
2. Solves issue [38942](https://github.com/pytorch/pytorch/issues/38942).
3. Fixes incorrect `find_unused_parameters_` passing like checking `outputs.empty()` or `unused_parameters_.empty()`.

Test Plan: 
1. Run `test/distributed/test_c10d.py` and make sure all tests pass.
2. A new test case `test_find_unused_parameters_when_unused_parameters_empty` is included. Old `reducer.cpp` was failing in that unit test because it was checking `find_unused_parameters_` by `unused_parameters_.empty()`. Current `reducer.cpp` passes this unit test.
3. Two test cases were failing `test_forward_backward_unused_parameters` and `test_forward_backward_optimizer` , because `find_unused_parameter_` of their `reducer` object was not set properly. I fixed that as well.

Tasks: T68705534

Tags: DDP

Differential Revision: [D22176231](https://www.internalfb.com/intern/diff/D22176231/)

[ghstack-poisoned]

zhaojuanmao commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -122,12 +122,15 @@ PyObject* c10d_init(PyObject* _unused) {
           py::arg("expect_sparse_gradients") = std::vector<std::vector<bool>>(),
           py::arg("bucket_bytes_cap") = ::c10d::kDefaultBucketBytesCap,
+          py::arg("find_unused_parameters") =
+              ::c10d::kDefaultFindUnusedParameters,
```
nit: it is OK to directly use False here

zhaojuanmao reviewed (commented):
overall lgtm

sinannasir committed:
Update on "Skip allreducing `local_used_maps_dev_` when `find_unused_param=False`"


Summary: 
1. In reducer.cpp, we have a new boolean `find_unused_param_` and its value is set in `Reducer::prepare_for_backward`.
If `!find_unused_param_`, then it avoids `allreduce(local_used_maps_dev_)`.
2. Solves issue [38942](https://github.com/pytorch/pytorch/issues/38942).
3. Fixes incorrect `find_unused_parameters_` passing like checking `outputs.empty()` or `unused_parameters_.empty()`.

Test Plan: 
1. Run `test/distributed/test_c10d.py` and make sure all tests pass.
2. A new test case `test_find_unused_parameters_when_unused_parameters_empty` is included. Old `reducer.cpp` was failing in that unit test because it was checking `find_unused_parameters_` by `unused_parameters_.empty()`. Current `reducer.cpp` passes this unit test.
3. Two test cases were failing `test_forward_backward_unused_parameters` and `test_forward_backward_optimizer` , because `find_unused_parameter_` of their `reducer` object was not set properly. I fixed that as well.

Tasks: T68705534

Tags: DDP

Differential Revision: [D22176231](https://www.internalfb.com/intern/diff/D22176231/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -128,36 +130,46 @@ Reducer::Reducer(
   }
 }
 
+// Note [Skip allreducing local_used_maps_dev]
```
Just curious, any reason for adding this note to the dtor?

sinannasir replied:
In Pritam's example, the notes were placed after the constructor. Thats why I put it there. 
https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/engine.cpp#L100. Example of referring to a note: https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/engine.cpp#L64.

mrshenli commented on test/distributed/test_c10d.py:
```diff
@@ -2389,6 +2388,62 @@ def run_and_verify_grad(model):
+            output.mean().backward()
+
+            # Now locally unused parameter should have grad updated on all ranks.
+            [self.assertIsNotNone(t_p.grad) for t_p in model.module.task_parameters()]
```
nit: I wonder if we should also verify that the grad values are as expected? We can do this by 

1. creating two local models, one's `forward` uses both `t0` and `t1`, and the other one's `forward` only uses `t1`
2. run forward/backward on them, and then compute the mean grads of these two models
3. Compare the mean grad with DDP's grad. 

I noticed the test above (`test_global_local_unused_params_grad`) didn't test the accuracy either. So feel free to do this in a followup PR.

sinannasir replied:
I noted this and do a followup soon.

mrshenli reviewed (approved):
LGTM! Thanks for working on this!

rohan-varma commented on test/distributed/test_c10d.py:
```diff
@@ -2389,6 +2388,62 @@ def run_and_verify_grad(model):
+        process_group = c10d.ProcessGroupGloo(store, self.rank, self.world_size)
+
+        # Test on CPU
+        cpu_model = DistributedDataParallel(
```
nit: In a follow up PR, It may be worth separating the GPU and CPU test so that (a portion of) this test can be run in CPU-only builds?

sinannasir replied:
Yes, I noted that.
